File: openbis_api/source/java/ch/ethz/sis/openbis/generic/shared/api/v3/dto/search/SampleSearchCriterion.java
/*
 * Copyright 2014 ETH Zuerich, CISD
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ch.ethz.sis.openbis.generic.shared.api.v3.dto.search;

import ch.systemsx.cisd.base.annotation.JsonObject;

/**
 * @author pkupczyk
 */
@JsonObject("SampleSearchCriterion")
public class SampleSearchCriterion extends AbstractEntitySearchCriterion
{

    private static final long serialVersionUID = 1L;

    private SampleSearchRelation relation;

    public SampleSearchCriterion()
    {
        this(SampleSearchRelation.SAMPLE);
    }

    SampleSearchCriterion(SampleSearchRelation relation)
    {
        this.relation = relation;
    }

    public SpaceSearchCriterion withSpace()
    {
        return with(new SpaceSearchCriterion());
    }

    public ExperimentSearchCriterion withExperiment()
    {
        return with(new ExperimentSearchCriterion());
    }

    public SampleParentsSearchCriterion withParents()
    {
        return with(new SampleParentsSearchCriterion());
    }

    public SampleChildrenSearchCriterion withChildren()
    {
        return with(new SampleChildrenSearchCriterion());
    }

    public SampleContainerSearchCriterion withContainer()
    {
        return with(new SampleContainerSearchCriterion());
    }

    public SampleSearchCriterion withOrOperator()
    {
        return (SampleSearchCriterion) withOperator(SearchOperator.OR);
    }

    public SampleSearchCriterion withAndOperator()
    {
        return (SampleSearchCriterion) withOperator(SearchOperator.AND);
    }

    public SampleSearchRelation getRelation()
    {
        return relation;
    }

    @Override
    protected SearchCriterionToStringBuilder createBuilder()
    {
        SearchCriterionToStringBuilder builder = super.createBuilder();
        builder.setName(relation.name());
        return builder;
    }

}

File: js-test/servers/common/core-plugins/tests/1/as/webapps/openbis-v3-api-test/html/models/dto/search/SampleSearchCriterion.js
/**
 * @author pkupczyk
 */
define([ "require", "support/stjs", "dto/search/AbstractEntitySearchCriterion", "dto/search/SampleSearchRelation", "dto/search/SpaceSearchCriterion", "dto/search/ExperimentSearchCriterion",
		"dto/search/SearchOperator" ], function(require, stjs, AbstractEntitySearchCriterion, SampleSearchRelation, SpaceSearchCriterion, ExperimentSearchCriterion, SearchOperator) {

	var SampleSearchCriterion = function() {
		AbstractEntitySearchCriterion.call(this);
		this.relation = SampleSearchRelation.SAMPLE;
	};
	stjs.extend(SampleSearchCriterion, AbstractEntitySearchCriterion, [ AbstractEntitySearchCriterion ], function(constructor, prototype) {
		prototype['@type'] = 'dto.search.SampleSearchCriterion';
		constructor.serialVersionUID = 1;
		prototype.relation = null;
		prototype.withSpace = function() {
			return this.addCriterion(new SpaceSearchCriterion());
		};
		prototype.withExperiment = function() {
			return this.addCriterion(new ExperimentSearchCriterion());
		};
		prototype.withParents = function() {
			return this.addCriterion(new SampleParentsSearchCriterion());
		};
		prototype.withChildren = function() {
			return this.addCriterion(new SampleChildrenSearchCriterion());
		};
		prototype.withContainer = function() {
			return this.addCriterion(new SampleContainerSearchCriterion());
		};
		prototype.withOrOperator = function() {
			return this.withOperator(SearchOperator.OR);
		};
		prototype.withAndOperator = function() {
			return this.withOperator(SearchOperator.AND);
		};
		prototype.getRelation = function() {
			return this.relation;
		};
		prototype.createBuilder = function() {
			var builder = AbstractEntitySearchCriterion.prototype.createBuilder.call(this);
			builder.setName(this.relation.name());
			return builder;
		};
	}, {
		relation : {
			name : "Enum",
			arguments : [ "SampleSearchRelation" ]
		},
		operator : {
			name : "Enum",
			arguments : [ "SearchOperator" ]
		},
		criteria : {
			name : "Collection",
			arguments : [ "ISearchCriterion" ]
		}
	});

	var SampleParentsSearchCriterion = function() {
		SampleSearchCriterion.call(this, SampleSearchRelation.PARENTS);
	};
	stjs.extend(SampleParentsSearchCriterion, SampleSearchCriterion, [ SampleSearchCriterion ], function(constructor, prototype) {
		prototype['@type'] = 'dto.search.SampleParentsSearchCriterion';
		constructor.serialVersionUID = 1;
	}, {
		relation : {
			name : "Enum",
			arguments : [ "SampleSearchRelation" ]
		},
		operator : {
			name : "Enum",
			arguments : [ "SearchOperator" ]
		},
		criteria : {
			name : "Collection",
			arguments : [ "ISearchCriterion" ]
		}
	});

	var SampleChildrenSearchCriterion = function() {
		SampleSearchCriterion.call(this, SampleSearchRelation.CHILDREN);
	};
	stjs.extend(SampleChildrenSearchCriterion, SampleSearchCriterion, [ SampleSearchCriterion ], function(constructor, prototype) {
		prototype['@type'] = 'dto.search.SampleChildrenSearchCriterion';
		constructor.serialVersionUID = 1;
	}, {
		relation : {
			name : "Enum",
			arguments : [ "SampleSearchRelation" ]
		},
		operator : {
			name : "Enum",
			arguments : [ "SearchOperator" ]
		},
		criteria : {
			name : "Collection",
			arguments : [ "ISearchCriterion" ]
		}
	});

	var SampleContainerSearchCriterion = function() {
		SampleSearchCriterion.call(this, SampleSearchRelation.CONTAINER);
	};
	stjs.extend(SampleContainerSearchCriterion, SampleSearchCriterion, [ SampleSearchCriterion ], function(constructor, prototype) {
		prototype['@type'] = 'dto.search.SampleContainerSearchCriterion';
		constructor.serialVersionUID = 1;
	}, {
		relation : {
			name : "Enum",
			arguments : [ "SampleSearchRelation" ]
		},
		operator : {
			name : "Enum",
			arguments : [ "SearchOperator" ]
		},
		criteria : {
			name : "Collection",
			arguments : [ "ISearchCriterion" ]
		}
	});

	return SampleSearchCriterion;
})